weasyl/two_factor_auth.py

```diff
@@ -0,0 +1,264 @@
+    tfa_secret = d.engine.scalar("""
+        SELECT twofa_secret
+        FROM login
+        WHERE userid = (%(userid)s)
+    """, userid=userid)
+    # Validate supplied 2FA response versus calculated current TOTP value.
+    totp = pyotp.TOTP(tfa_secret)
+    if totp.verify(tfa_response):
+        return True
+    # TOTP verification failed, check recovery code
+    elif is_recovery_code_valid(userid, tfa_response):
+        # Recovery code was valid, and consumed
+        return True
+    else:
+        # Received input was not a valid TOTP response, and was not a valid recovery code
+        return False
+
+
+def get_number_of_recovery_codes(userid):
+    """
+    Get and return the number of remaining recovery codes for `userid`.
+    
+    Parameters:
+        userid: The userid for which to check the count of recovery codes.
+
+    Returns:
+        An integer representing the number of remaining recovery codes.
+    """
+    return d.engine.scalar("""
+        SELECT COUNT(*)
+        FROM twofa_recovery_codes
+        WHERE userid = (%(userid)s)
+    """, userid=userid)
+
+
+def generate_recovery_codes(userid):
+    """
+    Generate a fresh set of 2FA recovery codes for a user.
+
+    Initializes or otherwise refreshes the 2FA recovery codes for a user,
+    by deleting all rows from the recovery table where the userid matches the
+    supplied userid, and generates fresh codes to supply to the user.
+
+    Parameters:
+        userid: The userid to create new recovery codes for.
+
+    Returns:
+        A set of recovery codes linked to the passed userid.
+    """
+    # First, purge existing recovery codes (if any).
+    d.engine.execute("""
+        DELETE FROM twofa_recovery_codes
+        WHERE userid = (%(userid)s);
+    """, userid=userid)
+    # Next, generate the recovery codes, up to the defined maximum value
+    tfa_recovery_codes = {security.generate_key(20)}
+    for i in range(0, _TFA_RECOVERY_CODES - 1):
+        tfa_recovery_codes |= {security.generate_key(20)}
+    # Then, insert the codes into the table
+    d.engine.execute("""
+        INSERT INTO twofa_recovery_codes (userid, recovery_code)
+        SELECT (%(userid)s), unnest( (%(tfa_recovery_codes)s) )
+    """, userid=userid, tfa_recovery_codes=list(tfa_recovery_codes))
+    # Finally, return the set of recovery codes to the calling function.
+    return tfa_recovery_codes
+
+
+def is_recovery_code_valid(userid, tfa_code):
+    """
+    Checks the recovery code table for a valid recovery code.
+
+    Determine if a supplied recovery code is present in the recovery code table
+    for a specified userid. If present, consume the code by deleting the record.
+
+    Parameters:
+        userid: The userid of the requesting user.
+        tfa_code: A candidate recovery code to check.
+
+    Returns: Boolean True if the code was valid and has been consumed, Boolean False, otherwise.
+    """
+    # Check to see if the provided code is valid, and consume if so
```

Comment:
Above this line: Ignore everything that is not length 20 (recovery code) here; don't even bother trying to hit the DB if it doesn't match the expected length; sanity check, even though ``verify()`` should successfully catch this (for extensibility, perhaps?).
```
if len(tfa_response) != 20:
    return False
```